src/analyzer/protocol/rdp/rdpeudp-analyzer.pac

```diff
@@ -0,0 +1,65 @@
+refine connection RDPEUDP_Conn += {
+	%member{
+		enum RDPEUDP_STATE {
+		        NEED_SYN	= 0x1,
+		        NEED_SYNACK	= 0x2,
+		        ESTABLISHED	= 0x3,
+		};
+		uint8 state_ = NEED_SYN;
+		bool is_rdpeudp2_ = false;
+	%}
+	function get_state(): uint8
+	%{
+		return state_;
+	%}
+
+        function proc_rdpeudp_syn(is_orig: bool, uFlags: uint16, snSourceAck: uint32): bool
+	%{
+		if (!is_orig) {
+			return false;
+		}
+                if ((uFlags & 0x01) == 0) {
+                        return false;
+                }
+		if (snSourceAck != 0xffffffff) {
+			return false;
+		}
+		if (uFlags >= 0x1000) {
+			is_rdpeudp2_ = true;
+		}
+                BifEvent::generate_rdpeudp_syn(bro_analyzer(), bro_analyzer()->Conn());
+		state_ = NEED_SYNACK;
+                return true;
+	%}
+
+        function proc_rdpeudp_synack(is_orig: bool, uFlags: uint16): bool
+	%{
+		if (is_orig) {
+			return false;
+		}
+
+		if ((uFlags & 0x05) == 0) {
+			return false;
+		}
+                BifEvent::generate_rdpeudp_synack(bro_analyzer(), bro_analyzer()->Conn());
+
+		if (is_rdpeudp2_) {
+			state_ = ESTABLISHED;
+	                BifEvent::generate_rdpeudp_established(bro_analyzer(), bro_analyzer()->Conn());
+		} else {
+			state_ = ESTABLISHED;
+        	        BifEvent::generate_rdpeudp_established(bro_analyzer(), bro_analyzer()->Conn());
+		}
+                return true;
+	%}
+
+        function proc_rdpeudp_ack(is_orig: bool, data: bytestring): bool
+	%{
+                BifEvent::generate_rdpeudp_data(bro_analyzer(),
+						bro_analyzer()->Conn(),
+						is_orig,
+						new StringVal(data.length(), (const char*) data.data())
+		);
```

Comment:
For all of the `BifEvent::generate_*` calls, it's often better to first check if an event handler exists:

```suggestion
		if ( rdpeudp_data )
	                BifEvent::generate_rdpeudp_data(bro_analyzer(),
							bro_analyzer()->Conn(),
							is_orig,
							new StringVal(data.length(), (const char*) data.data())
			);
```

It helps avoid doing unnecessary work, like in this case, it's easy to see we don't need to keep allocating memory for that `StringVal` if no one cares to consume it.

Speaking of performance/optimization another idea I wonder: if there's no handler for that "data" event maybe we can disable/remove the analyzer right after the rdpudp session gets established rather than continue to parse packets that no one will be interested in?

Reply:
I didn't know you could check for event handlers like that. Neat!

I think detaching the analyzer immediately after the session is established is reasonable. Once/if we figure out how to reassemble and pass TLS records to the ssl analyzer, I think the rdpedup analyzer can detach after the TLS handshake.

I imagine a flag, like SSL::disable_analyzer_after_detection but for rdpeudp, being available in scriptland.